src/Nrfutil/sandbox.ts

```diff
@@ -0,0 +1,446 @@
+    LogLevel,
+    LogMessage,
+    ModuleVersion,
+    NrfutilJson,
+    NrfUtilSettings,
+    Progress,
+    TaskEnd,
+    TaskProgress,
+} from './sandboxTypes';
+
+export type NrfutilSandboxType = ReturnType<typeof NrfutilSandbox>;
+
+const parseJsonBuffers = <T>(data: Buffer): T[] | undefined => {
+    const dataString = data.toString().trim();
+    if (!dataString.endsWith('}')) {
+        return undefined;
+    }
+    try {
+        return JSON.parse(`[${dataString.replaceAll('}\n{', '}\n,{')}]`) ?? [];
+    } catch {
+        return undefined;
+    }
+};
+
+const NrfutilSandbox = (
+    baseDir: string,
+    module: string,
+    version: string,
+    setting?: NrfUtilSettings
+) => {
+    const onLoggingHandlers: ((logging: LogMessage) => void)[] = [];
+    let logLevel: LogLevel = 'info';
+
+    const prepareEnv = () => {
+        const env = { ...process.env };
+        env.NRFUTIL_HOME = path.join(
+            baseDir,
+            'nrfutil-sandboxes',
+            module,
+            version
+        );
+        fs.mkdirSync(env.NRFUTIL_HOME, { recursive: true });
+
+        env.NRFUTIL_EXEC_PATH = path.join(env.NRFUTIL_HOME, 'bin');
+
+        if (setting?.bootstrapConfigUrl)
+            env.NRFUTIL_BOOTSTRAP_CONFIG_URL = setting.bootstrapConfigUrl;
+        if (setting?.bootstrapTarballPath)
+            env.NRFUTIL_BOOTSTRAP_TARBALL_PATH = setting.bootstrapTarballPath;
+        if (setting?.devicePluginsDirForceNrfdlLocation)
+            env.NRFUTIL_DEVICE_PLUGINS_DIR_FORCE_NRFDL_LOCATION =
+                setting.devicePluginsDirForceNrfdlLocation;
+        if (setting?.devicePluginsDirForceNrfutilLocation)
+            env.NRFUTIL_DEVICE_PLUGINS_DIR_FORCE_NRFUTIL_LIBDIR =
+                setting.devicePluginsDirForceNrfutilLocation;
+        if (setting?.ignoreMissingSubCommand)
+            env.NRFUTIL_IGNORE_MISSING_SUBCOMMAND =
+                setting.ignoreMissingSubCommand;
+        if (setting?.log) env.NRFUTIL_LOG = setting.log;
+        if (setting?.packageIndexUrl)
+            env.NRFUTIL_PACKAGE_INDEX_URL = setting.packageIndexUrl;
+
+        return env;
+    };
+
+    const processLoggingData = (data: NrfutilJson<unknown>) => {
+        if (data.type === 'log') {
+            onLoggingHandlers.forEach(onLogging =>
+                onLogging(data.data as unknown as LogMessage)
+            );
+            return true;
+        }
+
+        return false;
+    };
+
+    const commonParser = <Result>(
+        data: Buffer,
+        callbacks: {
+            onProgress?: (progress: Progress) => void;
+            onInfo?: (info: Result) => void;
+            onTaskEnd?: (taskEnd: TaskEnd<Result>) => void;
+            onLogging?: (logging: LogMessage) => void;
+        }
+    ): Buffer | undefined => {
+        const parsedData: NrfutilJson<Result>[] | undefined =
+            parseJsonBuffers(data);
+
+        if (!parsedData) {
+            return data;
+        }
+
+        parsedData.forEach(item => {
```

Comment:
All the magic casts for item.data should be handled by the types as proposed in the NrfutilJson comment